File: Trees&Heaps/Heap.java
package Trees;

import java.util.ArrayList;

public class Heap {
    public ArrayList<Integer> list;

    public Heap() {
        this.list = new ArrayList<>();
    }

    public static int parent(int index) {
        return (index - 1) / 2;
    }

    public static int left(int index) {
        return 2 * index + 1;
    }

    public static int right(int index) {
        return 2 * index + 2;
    }

    public void insert(int element) {
        list.add(element);
        upheap(list.size() - 1);
    }

    public void upheap(int index) {
        if (index == 0) {
            return;
        }
        int parentNode = parent(index);
        if (list.get(index) < list.get(parentNode)) {
            swap(index, parentNode);
            upheap(parentNode);
        }
    }

    private void swap(int first, int second) {
        int temp = list.get(first);
        list.set(first, list.get(second));
        list.set(second, temp);
    }

    private int remove() {
        if (list.size() == 0) {
            throw new Error("Heap is empty");
        }

        int root = list.get(0);
        int last = list.remove(list.size() - 1);

        if (list.size() != 0) {
            list.set(0, last);
            downheap(0);
        }

        return root;

    }

    private void downheap(int i) {
        int min = i;
        int leftChild = left(i);
        int rightChild = right(i);
        if (leftChild < list.size() && list.get(leftChild) < list.get(min)) {
            min = leftChild;
        }
        if (rightChild < list.size() && list.get(rightChild) < list.get(min)) {
            min = rightChild;
        }

        if (min != i) {
            swap(i, min);
            downheap(min);
        }
    }

    public static Heap populate(int[] elements) {
        Heap p = new Heap();
        for (int i : elements) {
            p.insert(i);
        }
        return p;
    }

    public static ArrayList<Integer> heapSort(int[] elements) {
        ArrayList<Integer> ans = new ArrayList<>();
        Heap p = populate(elements);
        while (!p.list.isEmpty()) {
            ans.add(p.remove());
        }

        return ans;
    }

    public static void main(String[] args) {
        System.out.println(heapSort(new int[]{13,9,25,12,4}));
    }
}

File: ADA/Heap.py
class Heap:
    def __init__(self):
        self.heap = []

    def parent(self, i):
        return (i - 1) // 2

    def left(self, i):
        return 2 * i + 1

    def right(self, i):
        return 2 * i + 2

    def swap(self, i, j):
        self.heap[i], self.heap[j] = self.heap[j], self.heap[i]

    def insert(self, value):
        self.heap.append(value)
        self.upheap(len(self.heap) - 1)

    def upheap(self, index):
        if index == 0:
            return
        p = self.parent(index)
        if self.heap[index] < self.heap[p]:
            self.swap(index, p)
            self.upheap(p)


    def remove(self):
        if not self.heap:
            raise Exception("Removing from an empty heap!")

        root = self.heap[0]
        last = self.heap.pop()

        if self.heap:
            self.heap[0] = last
            self.downheap(i=0)

        return root

    def downheap(self, i):
        min = i
        l = self.left(i)
        r = self.right(i)

        # Here we check (l < len(self.heap)) because it can out of index 
        # due to formula used : (l = 2*i + 1)
        if l < len(self.heap) and self.heap[l] < self.heap[min]:
            min = l
        if r < len(self.heap) and self.heap[r] < self.heap[min]:
            min = r

        if min != i:
            self.swap(i, j=min)
            self.downheap(min) 

    def heap_sort(self):
        result = []
        temp = self.heap[:]
        while self.heap:
            result.append(self.remove())
        self.heap = temp  # Restore heap after sorting
        return result


heap= Heap()
heap.insert(10)
heap.insert(9)
heap.insert(3)
heap.insert(7)
heap.insert(4)
heap.insert(8)
heap.insert(4)
print(heap.heap_sort())